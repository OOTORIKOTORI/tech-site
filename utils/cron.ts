// utils/cron.ts

/**
 * Cron 単一フィールドの内部表現
 * - values: 許容される数値リスト（昇順・重複なし）
 * - star: ワイルドカード（`*`）指定か
 */
export type CronField = { values: number[]; star: boolean }

/**
 * Cron 全体の内部表現（minute/hour/dom/month/dow）
 *
 * メモ:
 * - DOW は 0-6（0=Sun）。7 は非対応（エラー想定）。
 * - DOM×DOW は dowDomMode ('OR'|'AND') で切替。'*' は OR=unrestricted/AND=always-true。
 * - 名前トークン（JAN..DEC, SUN..SAT）は大小文字を無視。
 * - ステップ指定（例: asterisk/5, a-b/2）に対応。0 以下のステップは不正。
 * - 基準の秒・ミリ秒があるときは“次の分”に切り上げ。ちょうど分なら包括開始。
 */
export type CronSpec = {
  /** 秒（6フィールド版サポート: 省略時は 0 固定） */
  second: CronField
  minute: CronField
  hour: CronField
  dom: CronField
  month: CronField
  dow: CronField
  /**
   * DOM×DOW 判定モード: 'OR'|'AND'（未指定は 'OR'）
   */
  dowDomMode: 'OR' | 'AND'
  /** 入力が6フィールド（秒あり）だったかのフラグ */
  hasSeconds: boolean
}

function expandField(src: string, min: number, max: number): number[] {
  const parts = src.split(',')
  const set = new Set<number>()
  for (const partRaw of parts) {
    const [rawRange0, rawStep0] = partRaw.split('/')
    const rawRange = (rawRange0 ?? '*').trim()
    const stepNum = rawStep0 === undefined ? 1 : Number(rawStep0)
    if (!Number.isFinite(stepNum) || stepNum <= 0) {
      throw new Error('Step must be a positive integer')
    }
    const step = Math.floor(stepNum)

    if (rawRange === '*' && step > 1) {
      // ステップ計算: DOM では 1 から開始
      for (let v = min; v <= max; v += step) {
        set.add(v)
      }
    } else {
      const range = rawRange === '*' ? `${min}-${max}` : rawRange
      const [startStr, endStr] = range.includes('-') ? range.split('-') : [range, range]
      let start = Number(startStr)
      let end = Number(endStr ?? startStr)
      start = Math.max(min, Math.min(max, start))
      end = Math.max(min, Math.min(max, end))
      if (start > end) [start, end] = [end, start]
      for (let v = start; v <= end; v += step) set.add(v)
    }
  }
  return Array.from(set).sort((a, b) => a - b)
}

function makeField(src: string, min: number, max: number): CronField {
  return { values: expandField(src, min, max), star: src.trim() === '*' }
}

function makeDowField(src: string): CronField {
  const s = src.trim()
  if (s === '*') {
    return { values: [0, 1, 2, 3, 4, 5, 6], star: true }
  }
  // expandFieldで一旦数値化
  const vals = expandField(src, 0, 7)
  // 7が含まれていたらエラー
  if (vals.some(v => v === 7)) throw new Error('DOW supports 0-6 (0=Sun). 7 is not supported.')
  // 0..6以外は除外し、重複排除・昇順
  const uniq = Array.from(
    new Set(vals.map(v => (v >= 0 && v <= 6 ? v : undefined)).filter(v => v !== undefined))
  ).sort((a, b) => a - b)
  return { values: uniq as number[], star: false }
}

const MON_NAME: Record<string, number> = {
  JAN: 1,
  FEB: 2,
  MAR: 3,
  APR: 4,
  MAY: 5,
  JUN: 6,
  JUL: 7,
  AUG: 8,
  SEP: 9,
  OCT: 10,
  NOV: 11,
  DEC: 12,
}

const DOW_NAME: Record<string, number> = {
  SUN: 0,
  MON: 1,
  TUE: 2,
  WED: 3,
  THU: 4,
  FRI: 5,
  SAT: 6,
}

function mapNamedTokens(src: string, dict: Record<string, number>): string {
  return src
    .split(',')
    .map(tk => {
      const [rng, step] = tk.split('/')
      const r = (rng ?? '*').trim().toUpperCase()
      if (r === '*') return step ? `*/${step}` : '*'
      if (r === '') return `${r}/${step ?? ''}`.replace(/\/$/, '')
      if (r.includes('-')) {
        const pair = r.split('-')
        const a = pair[0] ?? ''
        const b = pair[1] ?? ''
        const aa = a && a in dict ? String(dict[a]) : a
        const bb = b && b in dict ? String(dict[b]) : b
        return `${aa}-${bb}${step ? '/' + step : ''}`
      } else {
        const v = r in dict ? String(dict[r]) : r
        return `${v}${step ? '/' + step : ''}`
      }
    })
    .join(',')
  // 名前トークンは必ず0..6にマップ
}

/**
 * Cron 式を解析して `CronSpec` に変換します。
 *
 * 仕様要点:
 * - フィールドは `minute hour dom month dow` の 5 つ。
 * - DOW: 0-6（0=Sun）。7 は非対応（エラー想定）。
 * - DOM×DOW は dowDomMode ('OR'|'AND') で切替。'*' は OR=unrestricted/AND=always-true。
 * - 名前トークン（JAN..DEC, SUN..SAT）は大小文字を無視。
 * - ステップ指定（例: asterisk/5, a-b/5）で 0 以下や NaN は不正。
 *
 * 例:
 * - asterisk/5 9-18 * * 1-5
 * - 0 0 1 JAN *
 */
export function parseCron(expr: string, opts?: { dowDomMode?: 'OR' | 'AND' }): CronSpec {
  const norm = normalizeAlias(expr.trim())
  const tokens = norm.split(/\s+/)
  if (tokens.length !== 5 && tokens.length !== 6) {
    throw new Error('Cron expression must have 5 or 6 fields')
  }
  const hasSeconds = tokens.length === 6
  const [s, m, h, dom, mon, dow] = hasSeconds
    ? (tokens as [string, string, string, string, string, string])
    : (['0', ...tokens] as [string, string, string, string, string, string])
  const second = makeField(s, 0, 59)
  const minute = makeField(m, 0, 59)
  const hour = makeField(h, 0, 23)
  const d = makeField(dom, 1, 31)
  const mo = /[A-Za-z]/.test(mon)
    ? makeField(mapNamedTokens(mon, MON_NAME), 1, 12)
    : makeField(mon, 1, 12)
  const dw = /[A-Za-z]/.test(dow) ? makeDowField(mapNamedTokens(dow, DOW_NAME)) : makeDowField(dow)
  const mode = opts?.dowDomMode === 'AND' ? 'AND' : 'OR'
  // INFOログ（テスト環境では抑制）
  if (
    typeof console !== 'undefined' &&
    console.info &&
    typeof process !== 'undefined' &&
    process.env &&
    process.env.NODE_ENV !== 'test' &&
    process.env.NODE_ENV !== 'production'
  ) {
    console.info(`[cron] dowDomMode: ${mode}`)
  }
  return { second, minute, hour, dom: d, month: mo, dow: dw, dowDomMode: mode, hasSeconds }
}

function partsInTZ(date: Date, tz: 'Asia/Tokyo' | 'UTC') {
  if (tz === 'UTC') {
    return {
      second: date.getUTCSeconds(),
      minute: date.getUTCMinutes(),
      hour: date.getUTCHours(),
      dom: date.getUTCDate(),
      month: date.getUTCMonth() + 1,
      dow: date.getUTCDay(),
    }
  } else {
    const jstMs = date.getTime() + 9 * 3600000
    const jst = new Date(jstMs)
    return {
      second: jst.getUTCSeconds(),
      minute: jst.getUTCMinutes(),
      hour: jst.getUTCHours(),
      dom: jst.getUTCDate(),
      month: jst.getUTCMonth() + 1,
      dow: jst.getUTCDay(),
    }
  }
}

function matchesParts(
  spec: CronSpec,
  p: { second: number; minute: number; hour: number; dom: number; month: number; dow: number }
): boolean {
  const inSet = (f: CronField, v: number) => f.star || f.values.includes(v)

  // 基本フィールドチェック
  if (
    !inSet(spec.second, p.second) ||
    !inSet(spec.minute, p.minute) ||
    !inSet(spec.hour, p.hour) ||
    !inSet(spec.month, p.month)
  ) {
    return false
  }

  // DOM×DOW 判定
  const mode = spec.dowDomMode === 'AND' ? 'AND' : 'OR'
  const domHit = inSet(spec.dom, p.dom)
  const dowHit = inSet(spec.dow, p.dow)

  if (mode === 'AND') {
    // AND モード: 両方が true (star は常時 true)
    return (spec.dom.star || domHit) && (spec.dow.star || dowHit)
  } else {
    // OR モード: どちらかが true または両方が star
    if (spec.dom.star && spec.dow.star) {
      return true // 両方が * なら常に true
    }
    if (spec.dom.star) {
      return dowHit // dom=* なら dow のみチェック
    }
    if (spec.dow.star) {
      return domHit // dow=* なら dom のみチェック
    }
    return domHit || dowHit // 通常の OR
  }
}

function buildDateInTZ(
  parts: { y: number; mon: number; dom: number; hour: number; min: number; sec: number },
  tz: 'Asia/Tokyo' | 'UTC'
): Date {
  const offsetMs = tz === 'Asia/Tokyo' ? 9 * 3600000 : 0
  const ms =
    Date.UTC(parts.y, parts.mon - 1, parts.dom, parts.hour, parts.min, parts.sec, 0) - offsetMs
  return new Date(ms)
}

function yearInTZ(date: Date, tz: 'Asia/Tokyo' | 'UTC'): number {
  if (tz === 'UTC') return date.getUTCFullYear()
  const jstMs = date.getTime() + 9 * 3600000
  return new Date(jstMs).getUTCFullYear()
}

function setInTZ(
  base: Date,
  tz: 'Asia/Tokyo' | 'UTC',
  patch: Partial<{
    y: number
    month: number
    dom: number
    hour: number
    minute: number
    second: number
  }>
): Date {
  const p = partsInTZ(base, tz)
  const y = yearInTZ(base, tz)
  return buildDateInTZ(
    {
      y: patch.y ?? y,
      mon: patch.month ?? p.month,
      dom: patch.dom ?? p.dom,
      hour: patch.hour ?? p.hour,
      min: patch.minute ?? p.minute,
      sec: patch.second ?? p.second,
    },
    tz
  )
}

// --- Timezone helpers (generic IANA TZ support) ---
function tzOffsetMinutes(date: Date, tz: string): number {
  // Derive tz offset by comparing formatted local parts vs actual epoch
  const fmt = new Intl.DateTimeFormat('en-US', {
    timeZone: tz,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
  })
  const parts = fmt.formatToParts(date)
  const get = (t: Intl.DateTimeFormatPartTypes) => parts.find(p => p.type === t)?.value || '0'
  const y = Number(get('year'))
  const mo = Number(get('month'))
  const da = Number(get('day'))
  const hh = Number(get('hour'))
  const mi = Number(get('minute'))
  const ss = Number(get('second'))
  const asUTC = Date.UTC(y, mo - 1, da, hh, mi, ss)
  const diffMs = asUTC - date.getTime()
  return Math.round(diffMs / 60000) // minutes
}

function partsInAnyTZ(date: Date, tz: string) {
  if (tz === 'UTC' || tz === 'Asia/Tokyo') return partsInTZ(date, tz as 'UTC' | 'Asia/Tokyo')
  const offMin = tzOffsetMinutes(date, tz)
  const local = new Date(date.getTime() + offMin * 60000)
  return {
    second: local.getUTCSeconds(),
    minute: local.getUTCMinutes(),
    hour: local.getUTCHours(),
    dom: local.getUTCDate(),
    month: local.getUTCMonth() + 1,
    dow: local.getUTCDay(),
  }
}

function yearInAnyTZ(date: Date, tz: string): number {
  if (tz === 'UTC' || tz === 'Asia/Tokyo') return yearInTZ(date, tz as 'UTC' | 'Asia/Tokyo')
  const offMin = tzOffsetMinutes(date, tz)
  const local = new Date(date.getTime() + offMin * 60000)
  return local.getUTCFullYear()
}

function buildDateInAnyTZ(
  parts: { y: number; mon: number; dom: number; hour: number; min: number; sec: number },
  tz: string
): Date {
  if (tz === 'UTC' || tz === 'Asia/Tokyo') return buildDateInTZ(parts, tz as 'UTC' | 'Asia/Tokyo')
  // Initial naive guess: as if tz offset were 0
  const guessMs0 = Date.UTC(parts.y, parts.mon - 1, parts.dom, parts.hour, parts.min, parts.sec, 0)
  // Compute offset at guess and adjust
  let off = tzOffsetMinutes(new Date(guessMs0), tz)
  let ms = guessMs0 - off * 60000
  // Re-evaluate once in case crossing DST boundary changed offset
  const off2 = tzOffsetMinutes(new Date(ms), tz)
  if (off2 !== off) {
    off = off2
    ms = guessMs0 - off * 60000
  }
  // As a safety, if parts don't match (DST gaps), nudge forward by 1h up to 2 steps
  for (let i = 0; i < 2; i++) {
    const p = partsInAnyTZ(new Date(ms), tz)
    if (
      p.month === parts.mon &&
      p.dom === parts.dom &&
      p.hour === parts.hour &&
      p.minute === parts.min &&
      p.second === parts.sec
    ) {
      break
    }
    ms += 3600000
  }
  return new Date(ms)
}

function setInAnyTZ(
  base: Date,
  tz: string,
  patch: Partial<{
    y: number
    month: number
    dom: number
    hour: number
    minute: number
    second: number
  }>
): Date {
  if (tz === 'UTC' || tz === 'Asia/Tokyo') return setInTZ(base, tz as 'UTC' | 'Asia/Tokyo', patch)
  const p = partsInAnyTZ(base, tz)
  const y = yearInAnyTZ(base, tz)
  return buildDateInAnyTZ(
    {
      y: patch.y ?? y,
      mon: patch.month ?? p.month,
      dom: patch.dom ?? p.dom,
      hour: patch.hour ?? p.hour,
      min: patch.minute ?? p.minute,
      sec: patch.second ?? p.second,
    },
    tz
  )
}

export function nextRuns(spec: CronSpec, baseFrom: Date, tz: string, count: number): Date[] {
  const out: Date[] = []
  let cur = new Date(baseFrom.getTime())
  if (spec.hasSeconds) {
    // 秒指定あり: ミリ秒があれば次の秒、なければその秒から包括開始
    if (cur.getMilliseconds() !== 0) {
      const p0 = partsInAnyTZ(cur, tz)
      cur = setInAnyTZ(cur, tz, { second: p0.second + 1 })
      cur.setUTCMilliseconds(0)
    } else {
      cur.setUTCMilliseconds(0)
    }
  } else {
    // 従来通り: 秒・ミリ秒があれば次の分に切り上げ、分ちょうどは包括
    if (cur.getSeconds() !== 0 || cur.getMilliseconds() !== 0) {
      const p0 = partsInAnyTZ(cur, tz)
      cur = setInAnyTZ(cur, tz, { minute: p0.minute + 1, second: 0 })
      cur.setUTCMilliseconds(0)
    } else {
      cur.setUTCSeconds(0, 0)
    }
  }

  const secFirst = spec.second.values[0] ?? 0
  const minFirst = spec.minute.values[0] ?? 0
  const hourFirst = spec.hour.values[0] ?? 0
  const guardLimit = 5 * 366
  let dayGuard = 0

  while (out.length < count && dayGuard < guardLimit) {
    // 月そろえ
    let p = partsInAnyTZ(cur, tz)
    if (!spec.month.star && !spec.month.values.includes(p.month)) {
      const nextMon = spec.month.values.find(v => v >= p.month) ?? spec.month.values[0]!
      let y = yearInAnyTZ(cur, tz)
      if (nextMon < p.month) y += 1
      cur = buildDateInAnyTZ(
        { y, mon: nextMon, dom: 1, hour: hourFirst, min: minFirst, sec: secFirst },
        tz
      )
      p = partsInAnyTZ(cur, tz)
    }

    // 日（DOM×DOW）: dowDomMode に応じて判定
    const inSet = (f: CronField, v: number) => f.star || f.values.includes(v)
    const domDowOk = (pp: { dom: number; dow: number }) => {
      // 実在日チェック: 日付が存在しない場合は false
      const pFull = partsInAnyTZ(cur, tz)
      const actualDate = new Date(yearInAnyTZ(cur, tz), pFull.month - 1, pp.dom)
      if (actualDate.getDate() !== pp.dom) return false

      const mode = spec.dowDomMode === 'AND' ? 'AND' : 'OR'
      const domHit = inSet(spec.dom, pp.dom)
      const dowHit = inSet(spec.dow, pp.dow)

      if (mode === 'AND') {
        return (spec.dom.star || domHit) && (spec.dow.star || dowHit)
      } else {
        // OR モード: どちらかが true または両方が star
        if (spec.dom.star && spec.dow.star) {
          return true // 両方が * なら常に true
        }
        if (spec.dom.star) {
          return dowHit // dom=* なら dow のみチェック
        }
        if (spec.dow.star) {
          return domHit // dow=* なら dom のみチェック
        }
        return domHit || dowHit // 通常の OR
      }
    }
    let ok = domDowOk(p)
    if (!ok) {
      cur = setInAnyTZ(cur, tz, { hour: hourFirst, minute: minFirst, second: secFirst })
      const maxSteps =
        spec.dow.star && !spec.dom.star ? 62 : spec.dom.star && !spec.dow.star ? 7 : 31
      let steps = 0
      while (!ok && steps < maxSteps && dayGuard < guardLimit) {
        const pp = partsInAnyTZ(cur, tz)
        cur = setInAnyTZ(cur, tz, {
          dom: pp.dom + 1,
          hour: hourFirst,
          minute: minFirst,
          second: secFirst,
        })
        dayGuard++
        steps++
        p = partsInAnyTZ(cur, tz)
        ok = domDowOk(p)
      }
      if (!ok) continue
    }

    // 時
    p = partsInAnyTZ(cur, tz)
    const nextHour = spec.hour.values.find(v => v >= p.hour)
    if (nextHour === undefined) {
      cur = setInAnyTZ(cur, tz, {
        dom: p.dom + 1,
        hour: hourFirst,
        minute: minFirst,
        second: secFirst,
      })
      dayGuard++
      continue
    }
    if (nextHour !== p.hour)
      cur = setInAnyTZ(cur, tz, { hour: nextHour, minute: minFirst, second: secFirst })

    // 分
    p = partsInAnyTZ(cur, tz)
    const nextMin = spec.minute.values.find(v => v >= p.minute)
    if (nextMin === undefined) {
      const nh = spec.hour.values.find(v => v >= p.hour + 1)
      if (nh !== undefined)
        cur = setInAnyTZ(cur, tz, { hour: nh, minute: minFirst, second: secFirst })
      else {
        cur = setInAnyTZ(cur, tz, {
          dom: p.dom + 1,
          hour: hourFirst,
          minute: minFirst,
          second: secFirst,
        })
        dayGuard++
      }
      continue
    }
    if (nextMin !== p.minute) cur = setInAnyTZ(cur, tz, { minute: nextMin, second: secFirst })
    p = partsInAnyTZ(cur, tz)

    // 秒（6フィールド対応）
    const nextSec = spec.second.values.find(v => v >= p.second)
    if (nextSec === undefined) {
      // 次の分へ
      const nm = spec.minute.values.find(v => v >= p.minute + 1)
      if (nm !== undefined) cur = setInAnyTZ(cur, tz, { minute: nm, second: secFirst })
      else {
        const nh = spec.hour.values.find(v => v >= p.hour + 1)
        if (nh !== undefined)
          cur = setInAnyTZ(cur, tz, { hour: nh, minute: minFirst, second: secFirst })
        else {
          cur = setInAnyTZ(cur, tz, {
            dom: p.dom + 1,
            hour: hourFirst,
            minute: minFirst,
            second: secFirst,
          })
          dayGuard++
        }
      }
      continue
    }
    if (nextSec !== p.second) cur = setInAnyTZ(cur, tz, { second: nextSec })
    p = partsInAnyTZ(cur, tz)

    if (matchesParts(spec, p)) {
      out.push(new Date(cur.getTime()))
      if (spec.hasSeconds) {
        // 次の秒へ
        cur = setInAnyTZ(cur, tz, { second: p.second + 1 })
      } else {
        // 次の分へ
        cur = setInAnyTZ(cur, tz, { minute: p.minute + 1, second: 0 })
      }
    } else {
      if (spec.hasSeconds) {
        cur = setInAnyTZ(cur, tz, { second: p.second + 1 })
      } else {
        cur = setInAnyTZ(cur, tz, { minute: p.minute + 1, second: 0 })
      }
    }
  }

  // 最終的に重複排除（getTimeで一意化）
  return Array.from(new Map(out.map(d => [d.getTime(), d])).values())
}

// --- エイリアスの正規化（@hourly など） ---
function normalizeAlias(expr: string): string {
  if (!expr.startsWith('@')) return expr
  const k = expr.toLowerCase()
  switch (k) {
    case '@yearly':
    case '@annually':
      return '0 0 1 1 *'
    case '@monthly':
      return '0 0 1 * *'
    case '@weekly':
      return '0 0 * * 0'
    case '@daily':
    case '@midnight':
      return '0 0 * * *'
    case '@hourly':
      return '0 * * * *'
    case '@minutely':
      return '* * * * *'
    default:
      throw new Error(`Unsupported alias: ${expr}`)
  }
}

// --- 簡易 humanize（日本語） ---
function toTimeLabel(spec: CronSpec): string {
  const mm = spec.minute.values
  const hh = spec.hour.values
  const ss = spec.second.values
  const isSingle = (arr: number[]) => arr.length === 1
  if (isSingle(hh) && isSingle(mm) && isSingle(ss)) {
    return `${hh[0]!}:${String(mm[0]!).padStart(2, '0')}:${String(ss[0]!).padStart(2, '0')}`
  }
  if (isSingle(hh) && isSingle(mm)) {
    return `${hh[0]!}:${String(mm[0]!).padStart(2, '0')}`
  }
  if (isSingle(hh) && mm.length > 1 && mm[0] === 0 && mm[mm.length - 1] === 55) {
    // 毎分相当
    return `${hh[0]!}時台 毎分`
  }
  // 分ステップの検出（*/n）
  const stepMin = detectStep(mm)
  if (isSingle(hh) && stepMin) {
    return `${hh[0]!}時台 の毎${stepMin}分`
  }
  if (hh.length > 1 && stepMin) {
    return `${hh[0]}-${hh[hh.length - 1]}時 の毎${stepMin}分`
  }
  return 'スケジュール'
}

function detectStep(values: number[]): number | null {
  if (values.length < 2) return null
  const step = values[1]! - values[0]!
  for (let i = 2; i < values.length; i++) {
    if (values[i]! - values[i - 1]! !== step) return null
  }
  return step > 0 ? step : null
}

function dowLabel(spec: CronSpec): string {
  const v = spec.dow
  if (v.star) return ''
  const vals = v.values
  if (vals.length === 5 && vals.join(',') === '1,2,3,4,5') return '平日'
  if (vals.length === 2 && vals.join(',') === '0,6') return '週末'
  if (vals.length === 1) return ['日曜', '月曜', '火曜', '水曜', '木曜', '金曜', '土曜'][vals[0]!]!
  return `${vals.map(n => ['日', '月', '火', '水', '木', '金', '土'][n] + '曜').join('・')}`
}

/**
 * 人間可読の日本語説明（タイムゾーン非依存の表現）。
 */
export function humanizeCron(spec: CronSpec): string
export function humanizeCron(spec: CronSpec, opts?: { tz?: string }): string
export function humanizeCron(expr: string): string
export function humanizeCron(expr: string, opts?: { tz?: string }): string
export function humanizeCron(specOrExpr: CronSpec | string, opts?: { tz?: string }): string {
  const spec = typeof specOrExpr === 'string' ? parseCron(specOrExpr) : specOrExpr
  // Keep API surface (opts.tz) without changing legacy output semantics
  // eslint-disable-next-line @typescript-eslint/no-unused-expressions
  void opts?.tz
  const dayPart = (() => {
    if (!spec.dom.star && spec.dow.star) {
      // 毎月X日
      const vs = spec.dom.values
      if (vs.length === 1) return `毎月${vs[0]}日`
      return `毎月 ${vs.join('・')}日`
    }
    const dlab = dowLabel(spec)
    if (dlab) return dlab
    return '毎日'
  })()
  const t = toTimeLabel(spec)
  // 既定は従来互換（tz指定があっても表示には影響させない：最小差分）
  return `${dayPart} ${t}`.trim()
}
