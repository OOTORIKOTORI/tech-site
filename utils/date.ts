// date utils: stable YYYY-MM-DD formatting without external deps
export function formatDateIso(d?: string): string {
  if (!d) return ''
  const m = /^\d{4}-\d{2}-\d{2}/.exec(d)
  if (m) return m[0]
  const dt = new Date(d)
  if (Number.isNaN(dt.getTime())) return ''
  const y = dt.getUTCFullYear()
  const m2 = String(dt.getUTCMonth() + 1).padStart(2, '0')
  const d2 = String(dt.getUTCDate()).padStart(2, '0')
  return `${y}-${m2}-${d2}`
}
