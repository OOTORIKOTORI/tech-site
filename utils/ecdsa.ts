// utils/ecdsa.ts
// ES256 (P-256) ECDSA signature conversion between DER and JOSE (base64url of 64-byte R||S)

function encodeBase64Url(bytes: Uint8Array): string {
  if (typeof Buffer !== 'undefined') {
    // eslint-disable-next-line no-undef
    const b64 = Buffer.from(bytes).toString('base64')
    return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '')
  }
  let bin = ''
  for (const b of bytes) bin += String.fromCharCode(b)
  if (typeof btoa !== 'function') throw new Error('btoa is not available')
  const b64 = btoa(bin)
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '')
}

function decodeBase64UrlToBytes(b64url: string): Uint8Array {
  const b64 = b64url.replace(/-/g, '+').replace(/_/g, '/') + '==='.slice((b64url.length + 3) % 4)
  if (typeof Buffer !== 'undefined') {
    // eslint-disable-next-line no-undef
    return new Uint8Array(Buffer.from(b64, 'base64'))
  }
  if (typeof atob !== 'function') throw new Error('atob is not available')
  const bin: string = atob(b64)
  const out = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i)
  return out
}

function trimLeadingZeros(buf: Uint8Array): Uint8Array {
  let i = 0
  while (i < buf.length && buf[i] === 0) i++
  return buf.slice(i)
}

function leftPad32(raw: Uint8Array): Uint8Array {
  if (raw.length > 32) throw new Error('Integer too large for P-256')
  const out = new Uint8Array(32)
  out.set(raw, 32 - raw.length)
  return out
}

function readLen(buf: Uint8Array, offset: number): { len: number; next: number } {
  if (offset >= buf.length) throw new Error('Invalid DER: truncated length')
  const first = buf[offset]
  if (first === undefined) throw new Error('Invalid DER: truncated length')
  offset += 1
  let len = first
  if (len < 0x80) {
    return { len, next: offset }
  }
  const numBytes = len & 0x7f
  if (numBytes === 0 || numBytes > 4) throw new Error('Invalid DER: length long form')
  if (offset + numBytes > buf.length) throw new Error('Invalid DER: truncated long length')
  len = 0
  for (let i = 0; i < numBytes; i++) {
    const v = buf[offset + i]
    if (v === undefined) throw new Error('Invalid DER: truncated long length')
    len = (len << 8) | v
  }
  return { len, next: offset + numBytes }
}

function writeLen(len: number): Uint8Array {
  if (len < 0x80) return new Uint8Array([len])
  const tmp: number[] = []
  let n = len
  while (n > 0) {
    tmp.push(n & 0xff)
    n >>= 8
  }
  tmp.reverse()
  return new Uint8Array([0x80 | tmp.length, ...tmp])
}

export function derToJoseB64Url(der: Uint8Array): string {
  if (der.length < 8) throw new Error('Invalid DER: too short')
  let off = 0
  if (der[off++] !== 0x30) throw new Error('Invalid DER: not a SEQUENCE')
  const seqLenInfo = readLen(der, off)
  const seqEnd = seqLenInfo.next + seqLenInfo.len
  off = seqLenInfo.next
  if (seqEnd !== der.length) {
    // Accept extra wrapping but usually should match
  }

  if (der[off++] !== 0x02) throw new Error('Invalid DER: expected INTEGER (r)')
  const rLenInfo = readLen(der, off)
  off = rLenInfo.next
  const rBytesFull = der.slice(off, off + rLenInfo.len)
  off += rLenInfo.len

  if (der[off++] !== 0x02) throw new Error('Invalid DER: expected INTEGER (s)')
  const sLenInfo = readLen(der, off)
  off = sLenInfo.next
  const sBytesFull = der.slice(off, off + sLenInfo.len)
  off += sLenInfo.len
  if (off !== seqEnd) {
    // trailing ignored but suspicious
  }

  const r = trimLeadingZeros(rBytesFull)
  const s = trimLeadingZeros(sBytesFull)
  if (r.length > 32 || s.length > 32) throw new Error('Invalid DER: integer too large')
  const r32 = leftPad32(r)
  const s32 = leftPad32(s)
  const jose = new Uint8Array(64)
  jose.set(r32, 0)
  jose.set(s32, 32)
  return encodeBase64Url(jose)
}

export function joseB64UrlToDer(joseSigB64Url: string): Uint8Array {
  const sig = decodeBase64UrlToBytes(joseSigB64Url)
  if (sig.length !== 64) throw new Error('Invalid JOSE signature: must be 64 bytes')
  const r = sig.slice(0, 32)
  const s = sig.slice(32, 64)
  const rTrim = trimLeadingZeros(r)
  const sTrim = trimLeadingZeros(s)

  const r0 = rTrim[0] ?? 0
  const s0 = sTrim[0] ?? 0
  const rNeedPad = rTrim.length > 0 && (r0 & 0x80) !== 0
  const sNeedPad = sTrim.length > 0 && (s0 & 0x80) !== 0
  const rDer = new Uint8Array((rNeedPad ? 1 : 0) + (rTrim.length || 1))
  const sDer = new Uint8Array((sNeedPad ? 1 : 0) + (sTrim.length || 1))
  if (rNeedPad) rDer[0] = 0x00
  if (sNeedPad) sDer[0] = 0x00
  if (rTrim.length === 0) rDer[rNeedPad ? 1 : 0] = 0
  else rDer.set(rTrim, rNeedPad ? 1 : 0)
  if (sTrim.length === 0) sDer[sNeedPad ? 1 : 0] = 0
  else sDer.set(sTrim, sNeedPad ? 1 : 0)

  const rHdr = new Uint8Array([0x02, ...Array.from(writeLen(rDer.length))])
  const sHdr = new Uint8Array([0x02, ...Array.from(writeLen(sDer.length))])
  const contentLen = rHdr.length + rDer.length + sHdr.length + sDer.length
  const seqHdr = new Uint8Array([0x30, ...Array.from(writeLen(contentLen))])

  const out = new Uint8Array(seqHdr.length + contentLen)
  let o = 0
  out.set(seqHdr, o)
  o += seqHdr.length
  out.set(rHdr, o)
  o += rHdr.length
  out.set(rDer, o)
  o += rDer.length
  out.set(sHdr, o)
  o += sHdr.length
  out.set(sDer, o)
  return out
}
