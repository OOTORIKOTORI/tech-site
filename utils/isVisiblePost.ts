// 公開判定関数: draft !== true && published !== false && internalタグを含まない（string型も解釈）
export function isVisiblePost(frontmatter: unknown): boolean {
  if (!frontmatter || typeof frontmatter !== 'object') return false
  const fm = frontmatter as {
    draft?: unknown
    published?: unknown
    tags?: unknown[] | string[]
  }
  const draft = fm.draft
  const published = fm.published
  const tags = Array.isArray(fm.tags) ? fm.tags : []

  const isDraft = draft === true || draft === 'true'
  const isUnpublished = published === false || published === 'false'
  const hasInternalTag = tags.some(tag => String(tag) === 'internal')

  return !isDraft && !isUnpublished && !hasInternalTag
}
