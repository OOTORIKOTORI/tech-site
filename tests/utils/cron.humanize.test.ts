import { describe, it, expect } from 'vitest'
import { parseCron, humanizeCron, nextRuns } from '../../utils/cron'

describe('cron humanize & 6-field/alias', () => {
  it('humanize: 平日 9:00 for 0 9 * * 1-5', () => {
    const spec = parseCron('0 9 * * 1-5')
    expect(humanizeCron(spec)).toMatchSnapshot()
  })

  it('alias: @hourly -> 0 * * * *', () => {
    const spec = parseCron('@hourly')
    expect(humanizeCron(spec)).toMatchSnapshot()
    // 次回計算（UTC基準）
    const base = new Date('2025-01-01T00:30:00.000Z')
    const runs = nextRuns(spec, base, 'UTC', 2).map(d => d.toISOString())
    expect(runs).toMatchSnapshot()
  })

  it('6-field: 30 0 9 * * 1 (Mon 09:00:30)', () => {
    const spec = parseCron('30 0 9 * * 1')
    expect(humanizeCron(spec)).toMatchSnapshot()
    const base = new Date('2025-01-06T08:59:50.000Z') // Mon 17:59:50 JST
    const next = nextRuns(spec, base, 'UTC', 3).map(d => d.toISOString())
    expect(next).toMatchSnapshot()
  })
})
