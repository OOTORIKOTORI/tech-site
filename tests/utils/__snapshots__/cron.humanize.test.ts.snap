// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cron humanize & 6-field/alias > 6-field: 30 0 9 * * 1 (Mon 09:00:30) 1`] = `"月曜 9:00:30"`;

exports[`cron humanize & 6-field/alias > 6-field: 30 0 9 * * 1 (Mon 09:00:30) 2`] = `
[
  "2025-01-06T09:00:30.000Z",
  "2025-01-13T09:00:30.000Z",
  "2025-01-20T09:00:30.000Z",
]
`;

exports[`cron humanize & 6-field/alias > alias: @hourly -> 0 * * * * 1`] = `"毎日 スケジュール"`;

exports[`cron humanize & 6-field/alias > alias: @hourly -> 0 * * * * 2`] = `
[
  "2025-01-01T01:00:00.000Z",
  "2025-01-01T02:00:00.000Z",
]
`;

exports[`cron humanize & 6-field/alias > humanize: 平日 9:00 for 0 9 * * 1-5 1`] = `"平日 9:00:00"`;
